packages/conf-libssl/conf-libssl.3/opam

```diff
@@ -0,0 +1,39 @@
+opam-version: "2.0"
+maintainer: "David Sheets <sheets@alum.mit.edu>"
+authors: ["The OpenSSL Project"]
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+homepage: "https://www.openssl.org/"
+license: "Apache-1.0"
+build: [
+  ["pkg-config" "--print-errors" "--exists" "openssl"]
+    {os != "freebsd" & os != "openbsd" & os != "netbsd" & # libssl is provided by base system on BSDs
+     os-distribution != "homebrew" & os-distribution != "macports" & os-distribution != "nixos"}
+  ["sh" "-c" "env \"PKG_CONFIG_PATH=$(brew --prefix openssl)/lib/pkgconfig:$PKG_CONFIG_PATH\" pkg-config --print-errors --exists openssl"] {os-distribution = "homebrew"}
+  ["sh" "-c" "env \"PKG_CONFIG_PATH=$HOME/.nix-profile/lib/pkgconfig:$PKG_CONFIG_PATH\" pkg-config --print-errors --exists openssl"] {os-distribution = "nixos"}
+  ["sh" "-c" "env \"PKG_CONFIG_PATH=/opt/local/lib/pkgconfig:$PKG_CONFIG_PATH\" pkg-config --print-errors --exists openssl"] {os-distribution = "macports"}
+]
+install: [
+  ["sh" "-c" "ln -s \"$(brew --prefix openssl)/lib/pkgconfig/openssl.pc\" '%{lib}%/pkgconfig/openssl.pc'"] {os-distribution = "homebrew"}
+  ["sh" "-c" "ln -s \"$HOME/.nix-profile/lib/pkgconfig/openssl.pc\" '%{lib}%/pkgconfig/openssl.pc'"] {os-distribution = "nixos"}
```

Comment:
Is this intended to be useful under `opam2nix`, or when running vanilla `opam` on NixOS?

If it's the former, I'm pretty sure it will fail - `$HOME` points to a dummy directory when building a derivation. But I also don't think it's necessary: the transitive dependency on `openssl` and pkg-config should result in [this setup hook](https://github.com/NixOS/nixpkgs/blob/ac6ac61e3f4255ec7ebd9fad36f410dddb0fde7a/pkgs/build-support/pkg-config-wrapper/setup-hook.sh) giving you the right PKG_CONFIG_PATH.

If you're talking about regular opam on NixOS, then I think this would work, if you have already run `nix-env -i openssl` or whatever first.

I see you're using `setenv` in conf-pkg-config, can you just add `setenv: [PKG_CONFIG_PATH += "$HOME/.nix-profile/lib/pkgconfig"]` there for nixos? That would be harmless if the path doesn't exist, and feels cleaner anyway imo.

Reply:
It's the later. I'm guessing the conf-* packages are ignored by `opam2nix`.

> I see you're using setenv in conf-pkg-config, can you just add setenv: [PKG_CONFIG_PATH += "$HOME/.nix-profile/lib/pkgconfig"] there for nixos? That would be harmless if the path doesn't exist, and feels cleaner anyway imo.

The thing here is that I cannot use shell variable inside `setenv` so sadly I don't think I can do that.

> If you're talking about regular opam on NixOS, then I think this would work, if you have already run nix-env -i openssl or whatever first.

Ok that seems to be what we want in nixos I'm guessing

Reply:
> It's the later. I'm guessing the conf-* packages are ignored by opam2nix.

Nope conf-* aren't treated specially. To `opam2nix`, they look like any other package with `depexts`.

> The thing here is that I cannot use shell variable inside setenv so sadly I don't think I can do that.

Ah, fair enough. FWIW Nix [discourages installing libraries profile-wide in `nix-env`](https://nixos.wiki/wiki/FAQ/I_installed_a_library_but_my_compiler_is_not_finding_it._Why%3F), so I'd suggest dropping this - it's going to make these `conf-*` packages work differently to other libraries, which will be a lot of work to apply uniformly and might be a debugging headache in the future. If you are running `opam` and end up depending on `ssl`, the recommendation is to do so within a `nix-shell -p openssl`, whether it's a literal invocation like that or putting it in a `shell.nix`.

If you do want to keep it, it probably needs to be more tolerant of errors:
 - add `-f` to `-ln` so it doesn't care if the destination path exists (it won't in opam2nix)
 - `mkdir -p '%{lib}%/pkgconfig'` first? You're doing a `mkdir` in `conf-pkg-config`, but the value of `%{lib}%` isn't the same between packages, is it? Even if they are the same path in `opam`, they're not in `opam2nix` so a `mkdir -p` would make this less brittle.